File: apps/gaussian/src/gaussian.cpp
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>

#include <assert.h>
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <time.h>
#include <omp.h>

#include "components.h"

//#define HEIGHT 512
//#define WIDTH 512
#define COMMENT "gaussian filter"
#define RGB_COMPONENT_COLOR 255

void gaussian(unsigned char* in, unsigned char* out, int HEIGHT, int WIDTH, int mul1, int mul2, int mul3, int mul4, int add1, int add2, int add3, int add4) {

    loop1: for (int i = 0; i < (HEIGHT/8)*8; i++) {
        loop2: for (int j = 0; j < (WIDTH/8)*8; j++) {
        // Set output to 0 if the 3x3 receptive field is out of bound.
            if ((i < 3) | (i > HEIGHT - 4) | (j < 3) | (j > WIDTH - 4)) {
                out[i * HEIGHT + j] = 0;
                continue;
            }
            int x1, x2, x3, x4, x5, x6, x7, x8, x9, x10, x11, x12, x13, x14, x15, x16, x17, x18, x19, x20,
                    x21, x22, x23, x24, x25, x26, x27, x28, x29, x30, x31, x32, x33, x34, x35, x36, x37, x38, x39, x40,
                    x41, x42, x43, x44, x45, x46, x47, x48, x49, x50, x51, x52, x53, x54, x55, x56, x57, x58, x59, x60, sum;

            x1 = mul[mul1](in[i * HEIGHT + j], 16);
            x2 = mul[mul2](in[(i - 1) * HEIGHT + (j - 1)], 5);
            x3 = mul[mul2](in[(i - 1) * HEIGHT + j], 5);
            x4 = mul[mul2](in[(i - 1) * HEIGHT + (j + 1)], 5);
            x5 = mul[mul2](in[i * HEIGHT + (j - 1)], 5);
            x6 = mul[mul2](in[i * HEIGHT + (j + 1)], 5);
            x7 = mul[mul2](in[(i + 1) * HEIGHT + (j - 1)], 5);
            x8 = mul[mul2](in[(i + 1) * HEIGHT + j], 5);
            x9 = mul[mul2](in[(i + 1) * HEIGHT + (j + 1)], 5);
            x10 = mul[mul3](in[(i - 2) * HEIGHT + (j - 1)], -3);
            x11 = mul[mul3](in[(i - 2) * HEIGHT + j], -3);
            x12 = mul[mul3](in[(i - 2) * HEIGHT + (j + 1)], -3);
            x13 = mul[mul3](in[(i - 1) * HEIGHT + (j - 2)], -3);
            x14 = mul[mul3](in[(i - 1) * HEIGHT + (j + 2)], -3);
            x15 = mul[mul3](in[i * HEIGHT + (j - 2)], -3);
            x16 = mul[mul3](in[i * HEIGHT + (j + 2)], -3);
            x17 = mul[mul3](in[(i + 1) * HEIGHT + (j - 2)], -3);
            x18 = mul[mul3](in[(i + 1) * HEIGHT + (j + 2)], -3);
            x19 = mul[mul3](in[(i + 2) * HEIGHT + (j - 1)], -3);
            x20 = mul[mul3](in[(i + 2) * HEIGHT + j], -3);
            x21 = mul[mul3](in[(i + 2) * HEIGHT + (j + 1)], -3);
            x22 = mul[mul4](in[(i - 2) * HEIGHT + (j - 2)], -2);
            x23 = mul[mul4](in[(i - 2) * HEIGHT + (j + 2)], -2);
            x24 = mul[mul4](in[(i + 2) * HEIGHT + (j - 2)], -2);
            x25 = mul[mul4](in[(i + 2) * HEIGHT + (j + 2)], -2);

            x26 = add[add1](x1, x2);
            x27 = add[add1](x3, x4);
            x28 = add[add1](x5, x6);
            x29 = add[add1](x7, x8);
            x30 = add[add1](x9, x10);
            x31 = add[add2](x11, x12);
            x32 = add[add2](x13, x14);
            x33 = add[add2](x15, x16);
            x34 = add[add2](x17, x18);
            x35 = add[add2](x19, x20);
            x36 = add[add2](x21, x22);
            x37 = add[add3](x23, x24);
            x38 = add[add3](x25, x26);
            x39 = add[add4](x27, x28);
            x40 = add[add4](x29, x30);
            x41 = add[add4](x31, x32);
            x42 = add[add4](x33, x34);
            x43 = add[add4](x35, x36);
            x44 = add[add4](x37, x38);
            x45 = add[add4](x39, x40);
            x46 = add[add4](x41, x42);
            x47 = add[add4](x43, x44);
            x48 = add[add4](x45, x46);
            x49 = add[add4](x47, x48);

            x50 = x49 - in[(i - 3) * HEIGHT + (j - 1)];
            x51 = x50 - in[(i - 3) * HEIGHT + j];
            x52 = x51 - in[(i - 3) * HEIGHT + (j + 1)];
            x53 = x52 - in[(i - 1) * HEIGHT + (j - 3)];
            x54 = x53 - in[(i - 1) * HEIGHT + (j + 3)];
            x55 = x54 - in[i * HEIGHT + (j - 3)];
            x56 = x55 - in[i * HEIGHT + (j + 3)];
            x57 = x56 - in[(i + 1) * HEIGHT + (j - 3)];
            x58 = x57 - in[(i + 1) * HEIGHT + (j + 3)];
            x59 = x58 - in[(i + 3) * HEIGHT + (j - 1)];
            x60 = x59 - in[(i + 3) * HEIGHT + j];
            sum = x60 - in[(i + 3) * HEIGHT + (j + 1)];

            sum = (sum < 0) ? 0 : sum;
            sum = (sum > 255) ? 255 : sum;
            out[i * HEIGHT + j] = (unsigned char)sum;
        }
    }
}

typedef struct {
    unsigned char red, green, blue;
} PPMPixel;

typedef struct {
    int x, y;
    PPMPixel *data;
} PPMImage;

static PPMImage *readPPM(const char *filename) {
    char buff[16];
    PPMImage *img;
    FILE *fp;
    int c, rgb_comp_color;
    fp = fopen(filename, "rb");
    if (!fp) {
        fprintf(stderr, "Unable to open file '%s'\n", filename);
        exit(1);
    }

    if (!fgets(buff, sizeof(buff), fp)) {
        perror(filename);
        exit(1);
    }

    if (buff[0] != 'P' || buff[1] != '6') {
        fprintf(stderr, "Invalid image format (must be 'P6')\n");
        exit(1);
    }

    img = (PPMImage *)malloc(sizeof(PPMImage));
    if (!img) {
        fprintf(stderr, "Unable to allocate memory\n");
        exit(1);
    }

    c = getc(fp);
    while (c == '#') {
        while (getc(fp) != '\n')
            ;
        c = getc(fp);
    }

    ungetc(c, fp);
    if (fscanf(fp, "%d %d", &img->x, &img->y) != 2) {
        fprintf(stderr, "Invalid image size (error loading '%s')\n", filename);
        exit(1);
    }

    if (fscanf(fp, "%d", &rgb_comp_color) != 1) {
        fprintf(stderr, "Invalid rgb component (error loading '%s')\n", filename);
        exit(1);
    }

    if (rgb_comp_color != RGB_COMPONENT_COLOR) {
        fprintf(stderr, "'%s' does not have 8-bits components\n", filename);
        exit(1);
    }

    while (fgetc(fp) != '\n')
        ;
    img->data = (PPMPixel *)malloc(img->x * img->y * sizeof(PPMPixel *));

    if (!img) {
        fprintf(stderr, "Unable to allocate memory\n");
        exit(1);
    }

    if (fread(img->data, 3 * img->x, img->y, fp) != img->y) {
        fprintf(stderr, "Error loading image '%s'\n", filename);
        exit(1);
    }

    fclose(fp);
    return img;
}

void writePPM(char* outname, unsigned char* img, char *filename, const char * kind, int height, int width) {
    FILE *output;
    
    char *token = strtok(filename, "p");
    strcpy(outname, token);

    strcat(outname, kind);
    strcat(outname, ".gauss.ppm");

    if ((output = fopen(outname, "w")) == NULL) {
        fprintf(stderr, "Error: could not open output file!\n");
        exit(1);
    }

    fprintf(output, "P2\n");
    fprintf(output, "%d %d\n", height, width);
    fprintf(output, "%d\n", RGB_COMPONENT_COLOR);

    for (int i = 0; i < height; i++) {
        for (int j = 0; j < width; j++)
            fprintf(output, "%d ", img[i * width + j]);
        fprintf(output, "\n");
    }
    fclose(output);
}

void init(unsigned char* buf, PPMImage * image) {
    for (int i = 0; i < image->y * image->x; i++) {
        buf[i] = image->data[i].red;
    }
}

// Output to stdout
void print(char **r, unsigned size) {
    for (int i = 0; i < size; i++) {
        printf(" %s", r[i]);
    }
    //printf("\n");
}

int main(int argc, char *argv[]) {
    FILE *input;
    char filename[30];
    unsigned size;
    char * kind;
    
    kind = argv[2];

    if ((input = fopen(argv[1], "r")) == NULL) {
        fprintf(stderr, "Error: could not open file\n");
        return 1;
    }

    // Read input filename
    fscanf(input, "%u", &size);
    
    char *filesList[size];
    char *inputList[size];
    
    for (int i = 0; i < size; i++) {
        fscanf(input, "%s\n", filename);
        inputList[i] = (char*) malloc (strlen(filename)+1);
        filesList[i] = (char*) malloc (strlen(filename)+8);
        strcpy(inputList[i],filename);
    }

    double t;
    double time_taken;
    t = omp_get_wtime();
    
#pragma omp parallel for shared(inputList, filesList)
    for (int i = 0; i < size; i++) {
        
        PPMImage *image = readPPM(inputList[i]);
        unsigned char *inputImage = (unsigned char *) malloc(sizeof(unsigned char) * image->y * image->x);
        unsigned char *outImage = (unsigned char *) malloc(sizeof(unsigned char) * image->y * image->x);

        init(inputImage, image);

        gaussian(inputImage, outImage, image->y, image->x, atoi(argv[3]), atoi(argv[4]), atoi(argv[5]), atoi(argv[6]), atoi(argv[7]), atoi(argv[8]), atoi(argv[9]), atoi(argv[10]));

        writePPM(filesList[i], outImage, inputList[i], kind, image->y, image->x);
        
        //printf("%s -> %s\n", inputList[i], filesList[i]);

        free(image);
        free(inputImage);
        free(outImage);
    }
    t = omp_get_wtime() - t;
    //time_taken = ((double) t) / CLOCKS_PER_SEC; // in seconds
    
    print(filesList, size);
    
    return EXIT_SUCCESS;
    
}

File: apps/gaussian/gaussian.py
from setuptools import setup
import scipy.stats as ss
import numpy as np
import evoapproxlib as eal
from PIL import Image
from skimage.metrics import structural_similarity as ssim
from skimage.metrics import mean_squared_error as mse
from skimage.metrics import peak_signal_noise_ratio as psnr
import cv2
import os
from multiprocessing import Pool, freeze_support
import subprocess

keys = ['PSNR', 'SSIM', 'MSE']

def baseline(pparams_dict, dict_app):
    r2 = subprocess.run([f"./apps/{dict_app['name']}/app", 
                         f"apps/{dict_app['name']}/{dict_app['training']}",
                         f"{pparams_dict['kind']}",
                         f"{pparams_dict['m0']}", f"{pparams_dict['m1']}", f"{pparams_dict['m2']}", f"{pparams_dict['m3']}",
                         f"{pparams_dict['a0']}", f"{pparams_dict['a1']}", f"{pparams_dict['a2']}", f"{pparams_dict['a3']}"], stdout=subprocess.PIPE)
    dict_app["metricerror"] = "SSIM"

def random_vector_norm(seed, size):
    x = np.arange(0, 65536)
    xU, xL = x + 0.5, x - 0.5 
    prob = ss.norm.cdf(xU, scale = 64) - ss.norm.cdf(xL, scale = 64)
    prob = prob / prob.sum() # normalize the probabilities so their sum is 1
    nums = np.random.default_rng(seed).choice(x, size, p = prob)
    return nums

def compute_error(original, approximate):
    psnr_value = []
    ssim_value = []
    mse_noise = []
    
    for i in range(len(original)):
        origin = cv2.imread(original[i])
        approx = cv2.imread(approximate[i])
        if psnr(origin, approx) != float("inf"):
            psnr_value.append(psnr(origin, approx))
        ssim_value.append(ssim(origin, approx, channel_axis=-1, gaussian_weights=True, sigma=1.5, use_sample_covariance=False, data_range=1.0))
        mse_noise.append(mse(origin, approx))
        
    if psnr_value: 
        psnr_value = np.mean(psnr_value)
    else: 
        psnr_value = 0
    
    return psnr_value, np.mean(ssim_value), np.mean(mse_noise)
    

def old_gaussian(params, **kwargs):
    input_images = kwargs.get('input')
    size = kwargs.get('size')
    kind = params['kind']

    values = []
    i = size[0]

    for case in input_images:
        if (kind != 'precise' or not os.path.exists('local-image/gaussian_' + kind + str(i) + '.png')):
            filtered_image = gaussian_kernel(params, case)
            filtered_image = filtered_image.astype(np.uint8)
            edge_image = Image.fromarray(filtered_image)
            edge_image.save('local-image/gaussian_' + kind + str(i) + '.png')
        values.append('local-image/gaussian_' + kind + str(i) + '.png')
        i += 1
        
    #print(values)
    return values

def image_run(i):
    
    filtered_image = gaussian_kernel(comps, input_images[i])
    filtered_image = filtered_image.astype(np.uint8)
    edge_image = Image.fromarray(filtered_image)
    edge_image.save('local-image/gaussian_' + kind + str(i+1) + '.png')
    
    return 'local-image/gaussian_' + kind + str(i+1) + '.png'
    

def gaussian_parallel(params, **kwargs):
    global comps
    comps = params
    global input_images
    input_images = kwargs.get('input')
    global size
    size = kwargs.get('size')
    global kind
    kind = params['kind']

    values = []
    
    if (kind != 'precise'):# or not os.path.exists('local-image/sobel_' + kind + str(i) + '.png')):
        with Pool() as pool:
            pool = Pool(processes=8)
            values = pool.map(image_run, range(0,size[1]))
    else:
        values = gaussian(params, **kwargs)

    #print(values)
    return values

def gaussian_kernel(params, input_image):

  [rows, columns] = np.shape(input_image)  # we need to know the shape of the input grayscale image
  filtered_image = np.zeros(shape=(rows, columns))  # initialization of the output image array (all elements are 0)

  # Now we "sweep" the image in both x and y directions and compute the output
  for i in range(rows):
      for j in range(columns):
          if ((i < 3) or (i > rows - 4) or (j < 3) or (j > columns - 4)):
              filtered_image[i][j] = 0
          else:
              x1 = params['m0'].calc(input_image[i][j], 16)
              x2 = params['m1'].calc(input_image[i - 1][j - 1], 5)
              x3 = params['m1'].calc(input_image[i - 1][j], 5)
              x4 = params['m1'].calc(input_image[i - 1][j + 1], 5)
              x5 = params['m1'].calc(input_image[i][j - 1], 5)
              x6 = params['m1'].calc(input_image[i][j + 1], 5)
              x7 = params['m1'].calc(input_image[i + 1][j - 1], 5)
              x8 = params['m1'].calc(input_image[i + 1][j], 5)
              x9 = params['m1'].calc(input_image[i + 1][j + 1], 5)
              x10 = params['m2'].calc(input_image[i - 2][j - 1], -3)
              x11 = params['m2'].calc(input_image[i - 2][j], -3)
              x12 = params['m2'].calc(input_image[i - 2][j + 1], -3)
              x13 = params['m2'].calc(input_image[i - 1][j - 2], -3)
              x14 = params['m2'].calc(input_image[i - 1][j + 2], -3)
              x15 = params['m2'].calc(input_image[i][j - 2], -3)
              x16 = params['m2'].calc(input_image[i][j + 2], -3)
              x17 = params['m2'].calc(input_image[i + 1][j - 2], -3)
              x18 = params['m2'].calc(input_image[i + 1][j + 2], -3)
              x19 = params['m2'].calc(input_image[i + 2][j - 1], -3)
              x20 = params['m2'].calc(input_image[i + 2][j], -3)
              x21 = params['m2'].calc(input_image[i + 2][j + 1], -3)
              x22 = params['m3'].calc(input_image[i - 2][j - 2], -2)
              x23 = params['m3'].calc(input_image[i - 2][j + 2], -2)
              x24 = params['m3'].calc(input_image[i + 2][j - 2], -2)
              x25 = params['m3'].calc(input_image[i + 2][j + 2], -2)
              
              x26 = params['a0'].calc(x1, x2)
              x27 = params['a0'].calc(x3, x4)
              x28 = params['a0'].calc(x5, x6)
              x29 = params['a0'].calc(x7, x8)
              x30 = params['a0'].calc(x9, x10)
              x31 = params['a1'].calc(x11, x12)
              x32 = params['a1'].calc(x13, x14)
              x33 = params['a1'].calc(x15, x16)
              x34 = params['a1'].calc(x17, x18)
              x35 = params['a1'].calc(x19, x20)
              x36 = params['a1'].calc(x21, x22)
              x37 = params['a2'].calc(x23, x24)
              x38 = params['a2'].calc(x25, x26)
              x39 = params['a3'].calc(x27, x28)
              x40 = params['a3'].calc(x29, x30)
              x41 = params['a3'].calc(x31, x32)
              x42 = params['a3'].calc(x33, x34)
              x43 = params['a3'].calc(x35, x36)
              x44 = params['a3'].calc(x37, x38)
              x45 = params['a3'].calc(x39, x40)
              x46 = params['a3'].calc(x41, x42)
              x47 = params['a3'].calc(x43, x44)
              x48 = params['a3'].calc(x45, x46)
              x49 = params['a3'].calc(x47, x48)
              
              x50 = x49 - input_image[i - 3][j - 1]
              x51 = x50 - input_image[i - 3][j]
              x52 = x51 - input_image[i - 3][j + 1]
              x53 = x52 - input_image[i - 1][j - 3]
              x54 = x53 - input_image[i - 1][j + 3]
              x55 = x54 - input_image[i][j - 3]
              x56 = x55 - input_image[i][j + 3]
              x57 = x56 - input_image[i + 1][j - 3]
              x58 = x57 - input_image[i + 1][j + 3]
              x59 = x58 - input_image[i + 3][j - 1]
              x60 = x59 - input_image[i + 3][j]
              out = x60 - input_image[i + 3][j + 1]
              
              if (out < 0):	out = 0;
              if (out > 255): out = 255 
              filtered_image[i][j] = out

  return filtered_image

def gaussian(params, inputfile):
    if (params['kind'] != 'precise'):
        r2 = subprocess.run(["./apps/gaussian/app", 
                            f"apps/gaussian/{inputfile}",
                            f"{params['kind']}",
                            f"{params['m0']}",
                            f"{params['m1']}",
                            f"{params['m2']}",
                            f"{params['m3']}",
                            f"{params['a0']}",
                            f"{params['a1']}",
                            f"{params['a2']}",
                            f"{params['a3']}"], 
        stdout=subprocess.PIPE)
        res = r2.stdout.decode("utf-8").split(' ')
        res.pop(0)

        return res
    else:
        files = []
        init = 0
        end = 0
        if inputfile == "training.in":
            init = 1
            end = 70
        else:
            init = 66
            end = 95
        for i in range(init, end + 1):
            files.append("local-image/512x512/" + str(i) + ".precise.gauss.ppm")
            
        return files

def get_inputs(size):
    inputs = []

    for case in range(size[0],size[1] + 1):
        image_file = "inputs/image-dataset/gray-512x512/" + str(case) + ".png"
        inputs.append(np.asarray(Image.open(image_file)))

    inputs = {
        'input': inputs,
        'size': size
    }

    return inputs
